elastica/timestepper/symplectic_steppers.py

```diff
@@ -26,201 +31,157 @@
-        last_kin_step = _steps_and_prefactors[-1][1]
+    def step_methods(self: SymplecticStepperProtocol) -> SteppersOperatorsType:
+        # Let the total number of steps for the Symplectic method
+        # be (2*n + 1) (for time-symmetry).
+        _steps: list[OperatorType] = self.get_steps()
+        # Prefac here is necessary because the linear-exponential integrator
+        # needs only the prefactor and not the dt.
+        _prefactors: list[OperatorType] = self.get_prefactors()
+        assert int(np.ceil(len(_steps) / 2)) == len(
+            _prefactors
+        ), f"{len(_steps)=}, {len(_prefactors)=}"
 
-        last_kin_step(TimeStepper, System, time, dt)
-        return time + last_kin_prefactor(TimeStepper, dt)
+        # Separate the kinematic and dynamic steps
+        _kinematic_steps: list[OperatorType] = _steps[::2]
+        _dynamic_steps: list[OperatorType] = _steps[1::2]
 
+        def no_operation(*args: Any) -> None:
+            pass
 
-class _SystemCollectionStepper:
-    """
-    Symplectic stepper collection class
-    """
+        return tuple(
+            zip_longest(
+                _prefactors,
+                _kinematic_steps,
+                _dynamic_steps,
+                fillvalue=no_operation,
+            )
+        )
 
+    @property
+    def n_stages(self: SymplecticStepperProtocol) -> int:
+        return len(self.steps_and_prefactors)
+
+    def step(
+        self: SymplecticStepperProtocol,
+        SystemCollection: SystemCollectionType,
+        time: np.floating,
+        dt: np.floating,
+    ) -> np.floating:
+        return SymplecticStepperMixin.do_step(self, self.steps_and_prefactors, SystemCollection, time, dt)  # type: ignore[attr-defined]
+
+    # TODO: Merge with .step method in the future.
+    # DEPRECATED: Use .step instead.
     @staticmethod
     def do_step(
-        TimeStepper,
-        _steps_and_prefactors,
-        SystemCollection,
-        time: np.float64,
-        dt: np.float64,
-    ):
+        TimeStepper: SymplecticStepperProtocol,
+        steps_and_prefactors: SteppersOperatorsType,
+        SystemCollection: SystemCollectionType,
+        time: np.floating,
+        dt: np.floating,
+    ) -> np.floating:
         """
         Function for doing symplectic stepper over the user defined rods (system).
 
-        Parameters
-        ----------
-        SystemCollection: rod object
-        time: float
-        dt: float
-
         Returns
         -------
+        time: float
+            The time after the integration step.
 
         """
-        for kin_prefactor, kin_step, dyn_step in _steps_and_prefactors[:-1]:
+        for kin_prefactor, kin_step, dyn_step in steps_and_prefactors[:-1]:
 
-            for system in SystemCollection._memory_blocks:
-                kin_step(TimeStepper, system, time, dt)
+            for system in SystemCollection.systems():
+                kin_step(system, time, dt)
```

Comment:
Lets revert this back to blocks, instead of systems.

Reply:
[499ae6d](https://github.com/GazzolaLab/PyElastica/pull/367/commits/499ae6deb16de3bf16cbdf462cb5b6c4f3f2c0cb)

I made `block_systems()` to iterate block, and kept `systems()` to iterate individual systems. Also modified documentation string.

Also refactored `_get_sys_idx_if_valid` to `get_system_index` as I think it is an important method to provide. Since `_systems` is no longer available to mixins, only way to find index is using `get_system_index` and can access by `__getitem__`.

You'll see a lot of changes like:
```diff
- sys_idx = self._get_sys_idx_if_valid(rod)
- system = self._systems[sys_idx]
+ sys_idx = self.get_system_index(rod)
+ system = self[sys_idx]
```

Now, system list is properly shielded.

Reply:
> ```diff
> + system = self[sys_idx]
> ```

Is the behavior same as current implementation in master?

Reply:
Yes